tests/integration/test.replication.js

```diff
@@ -2519,6 +2519,326 @@ adapters.forEach(function (adapters) {
+        // Do one replication, to not start from 0
+        writeStrange = true;
+        return source.replicate.to(dbs.name);
+      }).then(function (result) {
+        writeStrange = false;
+        // Verify that checkpoints are indeed mismatching:
+        should.exist(checkpoint);
+        var target = new PouchDB(dbs.name);
+        return PouchDB.utils.Promise.all([
+          target.get(checkpoint),
+          source.get(checkpoint)
+        ]);
+      }).then(function (res) {
+        // [0] = target checkpoint, [1] = source checkpoint
+        res[0].session_id.should.not.equal(res[1].session_id);
+
+        doc._rev = put.rev;
+        doc.count++;
+        return source.put(doc, {});
+      }).then(function (err, result) {
+        // Trigger the mismatch on the 2nd replication
+        mismatch = true;
+        return source.replicate.to(dbs.name);
+      });
+    });
+
+    it('#3999-4 should "upgrade" an old checkpoint', function() {
+
+      var secondRound = false;
+      var writeStrange = false;
+      var checkpoint;
+      var source = new PouchDB(dbs.remote);
+      var target = new PouchDB(dbs.name);
+
+      // 1. This is where we fake the mismatch:
+      var putter = function (doc) {
+
+        // We need the checkpoint id so we can inspect it later
+        if (/local/.test(doc._id)) {
+          checkpoint = doc._id;
+        }
+
+        var args = [].slice.call(arguments, 0);
+
+        // Write an old-style checkpoint on the first replication:
+        if (writeStrange && doc.last_seq === 1) {
+          var newDoc = {
+            _id: doc._id,
+            last_seq: doc.last_seq
+          };
+
+          args.shift();
+          args.unshift(newDoc);
+        }
+
+        if(this === source) {
+          return sourcePut.apply(this, args);
+        }
+
+        return targetPut.apply(this, args);
+      };
+
+      var sourcePut = source.put;
+      source.put = putter;
+      var targetPut =  target.put;
+      target.put = putter;
+
+      var changes = source.changes;
+      source.changes = function (opts) {
+        if (secondRound) {
+          // Test 1: Check that we read the old style local doc
+          // and didn't start from 0
+          opts.since.should.be.at.least(1);
+        }
+        return changes.apply(source, arguments);
+      };
+
+       var doc = { _id: '3', count: 0 };
+
+       return source.put({ _id: '4', count: 1 }, {}).then(function () {
+         writeStrange = true;
+         return source.replicate.to(target);
+       }).then(function() {
+         writeStrange = false;
+         // Verify that we have old checkpoints:
+         should.exist(checkpoint);
+         var target = new PouchDB(dbs.name);
+         return PouchDB.utils.Promise.all([
+           target.get(checkpoint),
+           source.get(checkpoint)
+         ]);
+       }).then(function (res) {
+        // [0] = target checkpoint, [1] = source checkpoint
+        res[0].last_seq.should.equal(res[1].last_seq);
```

Comment:
@daleharvey the `opts.since.should.not.equal(0)` trick works fine when i do it locally! thanks for that!
i'd like to do a `collate(res[0].last_seq, res[1].last_seq)` here, but i don;t know how to access pouchdb-collate so it works both on node and browser... do you have any trick for that?